File: scripts/agent/history_context.py
"""为 tarball 深析 Agent 提供同组织/相似项目的历史分析先验。"""

import os
from typing import Any

import requests


SUPABASE_URL = os.environ.get("SUPABASE_URL", "")
SUPABASE_SERVICE_ROLE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")
MAX_REFERENCES = 6


def _headers() -> dict[str, str]:
    return {
        "apikey": SUPABASE_SERVICE_ROLE_KEY,
        "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE_KEY}",
    }


def _get(table: str, params: dict[str, str]) -> list[dict[str, Any]]:
    response = requests.get(
        f"{SUPABASE_URL}/rest/v1/{table}",
        headers=_headers(),
        params=params,
        timeout=30,
    )
    response.raise_for_status()
    return response.json()


def _topic_similarity(left: list[str], right: list[str]) -> float:
    a = {item.lower() for item in left}
    b = {item.lower() for item in right}
    if not a or not b:
        return 0.0
    return len(a & b) / max(len(a), len(b))


def load_historical_context(owner: str, name: str) -> list[dict[str, Any]]:
    """历史记录只是待复核先验；数据库未配置或查询失败时安全降级为空。"""
    if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE_KEY:
        return []

    try:
        targets = _get("repositories", {
            "owner": f"eq.{owner}",
            "name": f"eq.{name}",
            "select": "id,owner,full_name,primary_language,topics",
            "limit": "1",
        })
        if not targets:
            return []
        target = targets[0]

        same_owner = _get("repositories", {
            "owner": f"eq.{owner}",
            "select": "id,owner,full_name,primary_language,topics,stars",
            "order": "stars.desc",
            "limit": "150",
        })
        similar_language: list[dict[str, Any]] = []
        if target.get("primary_language"):
            similar_language = _get("repositories", {
                "primary_language": f"eq.{target['primary_language']}",
                "select": "id,owner,full_name,primary_language,topics,stars",
                "order": "stars.desc",
                "limit": "250",
            })

        repositories = {row["id"]: row for row in same_owner + similar_language}
        repositories.pop(target["id"], None)
        if not repositories:
            return []

        ids = list(repositories)
        analysis_rows: list[dict[str, Any]] = []
        select = (
            "repository_id,tier,model,prompt_version,project_summary_cn,opportunity_verdict,"
            "opportunity_score,adaptation_points,recommended_approach,reasoning,analysis_details,analyzed_at,created_at"
        )
        for offset in range(0, len(ids), 100):
            chunk = ",".join(str(repo_id) for repo_id in ids[offset:offset + 100])
            analysis_rows.extend(_get("analysis", {
                "repository_id": f"in.({chunk})",
                "tier": "in.(2,3)",
                "select": select,
            }))

        best: dict[int, dict[str, Any]] = {}
        for row in analysis_rows:
            current = best.get(row["repository_id"])
            row_key = (row.get("tier", 0), row.get("analyzed_at") or row.get("created_at") or "")
            current_key = (
                current.get("tier", 0),
                current.get("analyzed_at") or current.get("created_at") or "",
            ) if current else (-1, "")
            if row_key > current_key:
                best[row["repository_id"]] = row

        ranked = []
        target_topics = target.get("topics") or []
        for repo_id, analysis in best.items():
            repo = repositories[repo_id]
            same_org = repo.get("owner", "").lower() == owner.lower()
            same_language = bool(
                target.get("primary_language")
                and repo.get("primary_language") == target.get("primary_language")
            )
            similarity = _topic_similarity(target_topics, repo.get("topics") or [])
            if not ((same_org and (same_language or similarity > 0)) or similarity >= 0.2):
                continue
            score = (
                (8 if same_org else 0)
                + similarity * 6
                + (1.5 if analysis.get("tier", 0) >= 3 else 0)
                + min(1, float(analysis.get("opportunity_score") or 0) / 100)
            )
            ranked.append((score, repo, analysis))

        ranked.sort(key=lambda item: (-item[0], item[1]["full_name"]))
        result = []
        for _, repo, analysis in ranked[:MAX_REFERENCES]:
            details = analysis.get("analysis_details") or {}
            result.append({
                "source_repo": repo["full_name"],
                "relation_clues": {
                    "same_organization": repo.get("owner", "").lower() == owner.lower(),
                    "language": repo.get("primary_language"),
                    "topics": repo.get("topics") or [],
                },
                "analysis_level": f"tier-{analysis.get('tier')} / {analysis.get('prompt_version')} / {analysis.get('model')}",
                "summary": analysis.get("project_summary_cn"),
                "verdict": analysis.get("opportunity_verdict"),
                "opportunities": (analysis.get("adaptation_points") or [])[:3],
                "recommended_approach": analysis.get("recommended_approach"),
                "decision": details.get("decision"),
                "reasoning_excerpt": (analysis.get("reasoning") or "")[:800],
            })
        return result
    except (requests.RequestException, KeyError, TypeError, ValueError):
        return []


def load_current_repository_context(owner: str, name: str) -> dict[str, Any]:
    """加载当前仓库的确定性支持事实和 tier-2 假设，供 tier-3 独立复核。"""
    if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE_KEY:
        return {}
    try:
        repositories = _get("repositories", {
            "owner": f"eq.{owner}",
            "name": f"eq.{name}",
            "select": "id,full_name,description,primary_language,topics,license,stars",
            "limit": "1",
        })
        if not repositories:
            return {}
        repo = repositories[0]
        signals = _get("harmony_signals", {
            "repository_id": f"eq.{repo['id']}",
            "select": "support_availability,support_provenance,support_coverage,support_confidence,support_evidence,ohpm_matched,ohpm_packages,gitcode_matched,gitcode_repo_url,deepwiki_scope",
            "limit": "1",
        })
        overrides = _get("harmony_overrides", {
            "repository_id": f"eq.{repo['id']}",
            "select": "state,note,marked_by,marked_at",
            "limit": "1",
        })
        tier2 = _get("analysis", {
            "repository_id": f"eq.{repo['id']}",
            "tier": "eq.2",
            "select": "model,prompt_version,project_summary_cn,opportunity_verdict,opportunity_score,adaptation_points,recommended_approach,reasoning,analysis_details,analyzed_at",
            "order": "analyzed_at.desc.nullslast",
            "limit": "1",
        })
        return {
            "repository": repo,
            "support_facts": signals[0] if signals else None,
            "manual_override": overrides[0] if overrides else None,
            "tier2_hypothesis": tier2[0] if tier2 else None,
        }
    except (requests.RequestException, KeyError, TypeError, ValueError):
        return {}

File: lib/llm/history-context.ts
import { getAdminClient } from '@/lib/supabase/admin';
import { stableHash } from '@/lib/hash';
import type { AdaptationPoint, AnalysisDetails, OpportunityVerdict } from '@/lib/types';

export interface HistoryTarget {
  id: number;
  owner: string;
  full_name: string;
  primary_language: string | null;
  topics: string[];
}

export interface HistoricalAnalysisReference {
  repository_id: number;
  full_name: string;
  owner: string;
  primary_language: string | null;
  topics: string[];
  tier: number;
  model: string;
  prompt_version: string;
  category: string | null;
  subcategory: string | null;
  project_summary_cn: string | null;
  opportunity_verdict: OpportunityVerdict | null;
  opportunity_score: number | null;
  opportunities: AdaptationPoint[];
  recommended_approach: string | null;
  reasoning_excerpt: string | null;
  decision: AnalysisDetails['decision'] | null;
  analyzed_at: string | null;
}

interface RepoMetadata {
  id: number;
  owner: string;
  full_name: string;
  primary_language: string | null;
  topics: string[] | null;
  stars: number;
}

interface AnalysisHistoryRow {
  repository_id: number;
  tier: number;
  model: string;
  prompt_version: string;
  category: string | null;
  subcategory: string | null;
  project_summary_cn: string | null;
  opportunity_verdict: OpportunityVerdict | null;
  opportunity_score: number | null;
  adaptation_points: AdaptationPoint[] | null;
  recommended_approach: string | null;
  reasoning: string | null;
  analysis_details: AnalysisDetails | null;
  analyzed_at: string | null;
  created_at: string;
}

const MAX_REFERENCES = 6;
const MAX_REPOSITORY_CANDIDATES = 1200;

function chunks<T>(values: T[], size: number): T[][] {
  const result: T[][] = [];
  for (let i = 0; i < values.length; i += size) result.push(values.slice(i, i + size));
  return result;
}

function topicSimilarity(left: string[], right: string[]): number {
  const a = new Set(left.map((item) => item.toLowerCase()));
  const b = new Set(right.map((item) => item.toLowerCase()));
  if (!a.size || !b.size) return 0;
  let overlap = 0;
  for (const topic of a) if (b.has(topic)) overlap++;
  return overlap / Math.max(a.size, b.size);
}

function relationScore(target: HistoryTarget, candidate: HistoricalAnalysisReference): number {
  const sameOwner = target.owner.toLowerCase() === candidate.owner.toLowerCase();
  const sameLanguage = Boolean(
    target.primary_language && candidate.primary_language === target.primary_language,
  );
  const topicScore = topicSimilarity(target.topics, candidate.topics);
  return (
    (sameOwner ? 8 : 0) +
    (sameLanguage ? 1.5 : 0) +
    topicScore * 6 +
    (candidate.tier >= 3 ? 1.5 : 0) +
    Math.min(1, (candidate.opportunity_score ?? 0) / 100)
  );
}

/**
 * 历史结论只作为可验证先验。优先同组织，其次选择语言与 topics 真正相似的项目；
 * 单纯“同为 TypeScript”不构成可复用关系。
 */
export function selectHistoricalReferences(
  target: HistoryTarget,
  candidates: HistoricalAnalysisReference[],
  limit = MAX_REFERENCES,
): HistoricalAnalysisReference[] {
  return candidates
    .filter((candidate) => candidate.repository_id !== target.id)
    .map((candidate) => ({
      candidate,
      sameOwner: target.owner.toLowerCase() === candidate.owner.toLowerCase(),
      sameLanguage: Boolean(target.primary_language && candidate.primary_language === target.primary_language),
      topics: topicSimilarity(target.topics, candidate.topics),
      score: relationScore(target, candidate),
    }))
    .filter(({ sameOwner, sameLanguage, topics }) => (sameOwner && (sameLanguage || topics > 0)) || topics >= 0.2)
    .sort((a, b) => b.score - a.score || b.candidate.tier - a.candidate.tier || a.candidate.full_name.localeCompare(b.candidate.full_name))
    .slice(0, limit)
    .map(({ candidate }) => candidate);
}

function preferAnalysis(next: AnalysisHistoryRow, current?: AnalysisHistoryRow): boolean {
  if (!current) return true;
  if (next.tier !== current.tier) return next.tier > current.tier;
  const nextTime = next.analyzed_at ?? next.created_at;
  const currentTime = current.analyzed_at ?? current.created_at;
  return nextTime > currentTime;
}

async function loadRelatedRepositories(targets: HistoryTarget[]): Promise<RepoMetadata[]> {
  const client = getAdminClient();
  const byId = new Map<number, RepoMetadata>();
  const owners = [...new Set(targets.map((target) => target.owner).filter(Boolean))];
  const languages = [...new Set(targets.map((target) => target.primary_language).filter((v): v is string => Boolean(v)))];

  for (const ownerChunk of chunks(owners, 30)) {
    const { data, error } = await client
      .from('repositories')
      .select('id,owner,full_name,primary_language,topics,stars')
      .in('owner', ownerChunk)
      .order('stars', { ascending: false })
      .limit(MAX_REPOSITORY_CANDIDATES);
    if (error) throw new Error(`加载同组织历史仓库失败:${error.message}`);
    for (const row of (data ?? []) as RepoMetadata[]) byId.set(row.id, row);
  }

  for (const languageChunk of chunks(languages, 12)) {
    const { data, error } = await client
      .from('repositories')
      .select('id,owner,full_name,primary_language,topics,stars')
      .in('primary_language', languageChunk)
      .order('stars', { ascending: false })
      .limit(MAX_REPOSITORY_CANDIDATES);
    if (error) throw new Error(`加载相似技术栈历史仓库失败:${error.message}`);
    for (const row of (data ?? []) as RepoMetadata[]) byId.set(row.id, row);
  }

  return [...byId.values()];
}

async function loadBestAnalyses(repositoryIds: number[]): Promise<Map<number, AnalysisHistoryRow>> {
  const client = getAdminClient();
  const best = new Map<number, AnalysisHistoryRow>();
  for (const idChunk of chunks(repositoryIds, 500)) {
    const { data, error } = await client
      .from('analysis')
      .select('repository_id,tier,model,prompt_version,category,subcategory,project_summary_cn,opportunity_verdict,opportunity_score,adaptation_points,recommended_approach,reasoning,analysis_details,analyzed_at,created_at')
      .in('repository_id', idChunk)
      .in('tier', [2, 3]);
    if (error) throw new Error(`加载历史深度分析失败:${error.message}`);
    for (const row of (data ?? []) as AnalysisHistoryRow[]) {
      if (preferAnalysis(row, best.get(row.repository_id))) best.set(row.repository_id, row);
    }
  }
  return best;
}

export async function loadHistoricalAnalysisContexts(
  targets: HistoryTarget[],
): Promise<Map<number, HistoricalAnalysisReference[]>> {
  const result = new Map<number, HistoricalAnalysisReference[]>();
  if (!targets.length) return result;
  const repositories = await loadRelatedRepositories(targets);
  const analyses = await loadBestAnalyses(repositories.map((repo) => repo.id));
  const references: HistoricalAnalysisReference[] = [];

  for (const repo of repositories) {
    const analysis = analyses.get(repo.id);
    if (!analysis) continue;
    references.push({
      repository_id: repo.id,
      full_name: repo.full_name,
      owner: repo.owner,
      primary_language: repo.primary_language,
      topics: repo.topics ?? [],
      tier: analysis.tier,
      model: analysis.model,
      prompt_version: analysis.prompt_version,
      category: analysis.category,
      subcategory: analysis.subcategory,
      project_summary_cn: analysis.project_summary_cn,
      opportunity_verdict: analysis.opportunity_verdict,
      opportunity_score: analysis.opportunity_score,
      opportunities: (analysis.adaptation_points ?? []).slice(0, 3),
      recommended_approach: analysis.recommended_approach,
      reasoning_excerpt: analysis.reasoning?.slice(0, 800) ?? null,
      decision: analysis.analysis_details?.decision ?? null,
      analyzed_at: analysis.analyzed_at,
    });
  }

  for (const target of targets) result.set(target.id, selectHistoricalReferences(target, references));
  return result;
}

export function historicalContextFingerprint(context?: HistoricalAnalysisReference[] | null): string | null {
  if (!context?.length) return null;
  return stableHash(context.map((item) => ({
    repo: item.full_name,
    tier: item.tier,
    version: item.prompt_version,
    model: item.model,
    language: item.primary_language,
    topics: [...item.topics].sort(),
    summary: item.project_summary_cn,
    verdict: item.opportunity_verdict,
    opportunities: item.opportunities.map((opportunity) => ({
      area: opportunity.area,
      integration_form: opportunity.integration_form,
      harmony_value: opportunity.harmony_value,
      target_devices: opportunity.target_devices,
    })),
    approach: item.recommended_approach,
    decision: item.decision,
  })));
}

export function formatHistoricalAnalysisContext(context?: HistoricalAnalysisReference[] | null): string {
  if (!context?.length) return '(没有足够相似且已有深度分析的历史项目)';
  return context.map((item) => JSON.stringify({
    source_repo: item.full_name,
    relation_clues: { owner: item.owner, language: item.primary_language, topics: item.topics },
    analysis_level: `tier-${item.tier} / ${item.prompt_version} / ${item.model}`,
    summary: item.project_summary_cn,
    verdict: item.opportunity_verdict,
    score: item.opportunity_score,
    reusable_patterns_only: item.opportunities.map((opportunity) => ({
      source_repo: item.full_name,
      area: opportunity.area,
      integration_form: opportunity.integration_form,
      harmony_value: opportunity.harmony_value,
      target_devices: opportunity.target_devices,
      validation_lessons: opportunity.validation_questions,
    })),
    source_repo_approach: item.recommended_approach,
    decision: item.decision,
    evidence_namespace: `上述内容全部属于 ${item.full_name}；其中没有任何路径可作为当前仓库证据`,
  }, null, 2)).join('\n\n');
}
